---
title: "seurat_cc_deg_calc"
output: html_document
date: "2022-08-24"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(SeuratData)

# library(devtools)
# devtools::install_github('mojaveazure/seurat-disk')
library(SeuratDisk)

# devtools::install_github("theislab/zellkonverter")
library(zellkonverter)

library(dplyr)

library(utils)
```

# Calculating Cell Cycle/Differential Gene Expression with Seurat

This is an R markdown notebook created to calculate cell cycle and differential gene expression scores through Seurat. It takes a .h5ad (ANNDATA) object as input.

## STEP 0: Prepare Env and Get Data
#### set global parameters
```{r}
main_dir <- "~/Coding/R/scrnaseq/"
setwd("~/Coding/R/scrnaseq")
in_path <- "D://Data//setd1a//setd1a_b4.log1p_liger_singleR"

cc_score_thresh <- 50
desired_cc_cell_types <- c("RG", "IPC")

deg_thresh <- 100
```

#### convert desired .h5ad into a Seurat object using `zellkonverter` package
```{r}
setd1a_sce <- readH5AD(paste(in_path, ".h5ad", sep = ""), verbose = TRUE)
setd1a <- as.Seurat(setd1a_sce, counts = "X", data = NULL)
```

#### initialise and preprocess the new Seurat object
``` {r}
setd1a <- NormalizeData(setd1a)
setd1a <- FindVariableFeatures(setd1a, selection.method = "vst")
setd1a <- ScaleData(setd1a, features = rownames(setd1a))

setd1a <- RunPCA(setd1a, features = VariableFeatures(setd1a), ndims.print = 6:10, nfeatures.print = 10)
```

## STEP 1: Calculate Cell Cycle Score

#### determine the desired cell types to calculate scores for
There are two parameters that can be changed to conduct a slightly different analysis:
* `cc_score_thresh`: determines how many occurrences a cell type needs to be included in analysis (default 50)
* `desired_cc_cell_types`: determines which types of cells (e.g. RG, IPC, EN, IN, etc.) should be included in analysis
``` {r}
cell_types <- c()

cell_type_all <- as.data.frame(setd1a$nowakowski.noglyc)

cell_type_counts <- cell_type_all %>% count(setd1a$nowakowski.noglyc)
cell_type_counts <- cell_type_counts[cell_type_counts$n > cc_score_thresh, ]
cell_type_distinct <- cell_type_counts %>% select("setd1a$nowakowski.noglyc")

for (i in cell_type_distinct$`setd1a$nowakowski.noglyc`) {
  desired = FALSE
  
  for (j in desired_cc_cell_types) {
    if (grepl(j, i, fixed = TRUE) == TRUE) {
      desired = TRUE
    }
  }
  if (desired == TRUE) {
    cell_types <- c(cell_types, i)
  }
}
```

#### initialise an empty master cc score dataframe and empty out dir
``` {r}
master_df <- data.frame()

dir.create(file.path("cc_out"), showWarnings = FALSE)
```


#### calculate cc scores
``` {r}
for (i in cell_types) {
  # cc markers loaded with Seurat
  s.genes <- cc.genes$s.genes
  g2m.genes <- cc.genes$g2m.genes
  
  # subset data 
  setd1a_sub <- subset(x = setd1a, subset = nowakowski.noglyc == i)
  
  # run cell cycle score analysis
  setd1a_sub <- CellCycleScoring(setd1a_sub, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
  
  # save plots/data
  pdf(file = paste(main_dir, "/cc_out/", i, "_cc_score_RidgePlot.pdf", sep = ""))
  print(
    RidgePlot(setd1a_sub, features = c("PCNA", "TOP2A", "MCM6", "MKI67"), ncol = 2)
  )
  dev.off()
  
  pdf(file = paste(main_dir, "/cc_out/", i, "_cc_score_DimPlot.pdf", sep = ""))
  print(
    DimPlot(setd1a_sub)
  )
  dev.off()
  
  df <- as.data.frame(setd1a_sub[[]])
  master_df <- master_df %>% bind_rows(df)
 
  write.csv(df, paste(main_dir, "/cc_out/", i, "_cc_score_data.csv", sep = ""), row.names = TRUE)
}
```

#### save master cc score df to a csv
``` {r}
write.csv(master_df, paste(main_dir, "/cc_out/all_cc_score_data.csv", sep = ""), row.names = TRUE)
```

#### run KS tests and generate a plot
``` {r}
# install.packages("stringr")
library(stringr)

# rg_s_ct3 <- read.table(file = "clipboard",
#                       sep = "\t", header = TRUE)
rg_s_ct3 <- rg_s_ct3$S.Score

# rg_s_fs3 <- read.table(file = "clipboard",
#                       sep = "\t", header = TRUE)
rg_s_fs3 <- rg_s_fs3$S.Score
group <- c(rep("rg_s_ct3", length(rg_s_ct3)), rep("rg_s_fs3", length(rg_s_fs3)))

dat <- data.frame(KSD = c(rg_s_ct3, rg_s_fs3), group = group)
dat <- dat %>% mutate(group = str_replace(group, "rg_s_ct3", "WT"))
dat <- dat %>% mutate(group = str_replace(group, "rg_s_fs3", "FS"))

ct_ecdf <- ecdf(rg_s_ct3)
fs_ecdf <- ecdf(rg_s_fs3)

# TESTING: KS-test
ks_r <- ks.test(rg_s_ct3, rg_s_fs3, 3, 2)
ks_text <- paste("KS TEST w/ D-val = ", format(ks_r[1], digits = 2), ", pval = ", format(ks_r[2], digits = 2), sep = "")

minMax <- seq(min(rg_s_ct3, rg_s_fs3), max(rg_s_ct3, rg_s_fs3), length.out = length(rg_s_ct3))
x0 <- minMax[which(abs(ct_ecdf(minMax) - fs_ecdf(minMax)) == max(abs(ct_ecdf(minMax) - fs_ecdf(minMax))))] 
y0 <- ct_ecdf(x0) 
y1 <- fs_ecdf(x0) 

p <- ggplot(dat, aes(x = KSD, group = group, color = group)) +
  stat_ecdf(size=1) +
  theme_bw(base_size = 28) +
  theme(legend.position ="top") +
  xlab("Cell Cycle S-Score") +
  ylab("ECDF") +
  #geom_line(size=1) +
  geom_segment(aes(x = x0[1], y = y0[1], xend = x0[1], yend = y1[1]),
               linetype = "dashed", color = "red") +
  geom_point(aes(x = x0[1] , y= y0[1]), color="red", size=8) +
  geom_point(aes(x = x0[1] , y= y1[1]), color="red", size=8) +
  ggtitle(ks_text) +
  theme(legend.title=element_blank()) +
  theme(axis.line = element_line(color='black'),
        plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank()) +
  scale_color_manual(breaks = c("WT", "FS"),
                     values = c("black", "red"))

ggsave(plot = p, filename = "rg_s_score.pdf", path = "~/Coding/R/scrnaseq/cc_out/",
       width = 10, height = 8, units = "in")
```

## STEP 2: calculate differential gene scores

#### determine the desired cell types to calculate scores for
There is one parameter that can be changed to conduct a slightly different analysis:
* `deg_thresh`: determines how many occurrences a cell type needs to be included in analysis (default 100)
``` {r}
cell_types <- c()

cell_type_all <- as.data.frame(setd1a$nowakowski.noglyc)

cell_type_counts <- cell_type_all %>% count(setd1a$nowakowski.noglyc)
cell_type_counts <- cell_type_counts[cell_type_counts$n > 100, ]
cell_type_distinct <- cell_type_counts %>% select("setd1a$nowakowski.noglyc")
```

#### initialise empty out dir
``` {r} 
dir.create(file.path("deg_out"), showWarnings = FALSE)
```

#### calculate diff gene exp scores with MAST

NOTE: When calculating diff gene exp scores, it is necessary to set the identities of the object to the `sampleID` column in order for the comparisons between different samples to be made. The identities can be changed as desired by using `Idents(object = [your Seurat object]) <- [desired data column]`.

``` {r}
for (i in cell_type_distinct$`setd1a$nowakowski.noglyc`) {
  
  setd1a_sub <- subset(x = setd1a, subset = nowakowski.noglyc == i)
  
  # set identities to compare between samples
  Idents(object = setd1a_sub) <- setd1a_sub$sampleID
  Idents(object = setd1a_sub)
  
  # install.packages("BiocManager")
  
  # run these if first time on local pc
  # BiocManager::install("limma")
  # BiocManager::install("MAST")
  # BiocManager::install("DESeq2")
  
  # save MAST score
  fs3_cntrl3_mast <- FindMarkers(setd1a_sub, ident.1 = "FS3", ident.2 = "CT3",
                                 test.use = "MAST", min.cells.group = 1,
                                 min.cells.feature = 1, min.pct = 0, 
                                 logfc.threshold = 0, only.pos = FALSE)
  write.csv(fs3_cntrl3_mast, paste(main_dir, "/deg_out/", "EN-V1", "_fs3_cntrl3_mast.csv", sep = ""), row.names = TRUE)
} 
```

#### calculate diff gene exp scores with DESeq2
WARNING: DESeq2 requires integral datapoints, and may throw an error if there are non-integers present.
``` {r}
for (i in cell_type_distinct$`setd1a$nowakowski.noglyc`) {
  
  setd1a_sub <- subset(x = setd1a, subset = nowakowski.noglyc == i)
  
  # set identities to compare between samples
  Idents(object = setd1a_sub) <- setd1a_sub$sampleID
  Idents(object = setd1a_sub)wwde
  
  # install.packages("BiocManager")
  
  # run these if first time on local pc
  # BiocManager::install("limma")
  # BiocManager::install("MAST")
  # BiocManager::install("DESeq2")
  
  # save MAST score
  fs3_cntrl3_deseq2 <- FindMarkers(setd1a_sub, ident.1 = "FS3", ident.2 = "CT3", test.use = "DESeq2")
  write.csv(fs3_cntrl3_deseq2, paste(main_dir, "/deg_out/", i, "_fs3_cntrl3_deseq2.csv", sep = ""), row.names = FALSE)
} 
```